### PR TITLE
[CodeGen][WinEH] Update saved esp for inlined inallocas

### PR DESCRIPTION
**Pull request by MuellerMP**
This fixes issue #116583

When inalloca calls are inlined the static stack pointer saving prolog of X86WinEHState breaks due to dynamic allocas. 
In this case we need to update the saved esp for every inalloca and for every stackrestore also related to inalloca.

### CODE REVIEW

File: llvm/lib/Target/X86/X86WinEHState.cpp [hunk 1]
```diff
@@ -774,3 +778,27 @@ void WinEHStatePass::insertStateNumberStore(Instruction *IP, int State) {
                                               RegNode, StateFieldIndex);
   Builder.CreateStore(Builder.getInt32(State), StateField);
 }
+
+void WinEHStatePass::updateEspForInAllocas(Function &F) {
+  for (BasicBlock &BB : F) {
+    for (Instruction &I : BB) {
+      if (auto *Alloca = dyn_cast<AllocaInst>(&I)) {
+        if (!Alloca->isUsedWithInAlloca())
```
Comment (rnk) [inline llvm/lib/Target/X86/X86WinEHState.cpp:786]: I think what you really want to do here is update the saved ESP after every stack pointer update, which means every dynamic alloca, and every stackrestore operation.

It's only correct to exclude inalloca allocas because it's generally impossible to recover from exceptions inside of a call expression, although it occurs to me that statement expressions make this possible (Ew).

For that reason, and given the general disrepair and non-performance criticality of 32-bit Windows code quality, I would recommend using `Alloca->isStaticAlloca()` as the criteria here.

**Review by rnk — COMMENTED** [review]
Thanks, this is not a fun bug.

File: llvm/lib/Target/X86/X86WinEHState.cpp (hunk 1 above)
Comment (MuellerMP) [reply llvm/lib/Target/X86/X86WinEHState.cpp]: updated

First of all thanks for the review.


Also added another test case for dynamic allocas.
I was unsure if this was actually a case we have to handle because my test cases never seemd to generate one.
But I see now that you can just use the c alloca api directly to generate one.

Regarding the inalloca case:
Statement expressions are not the only way to introduce this hazard - inlining does as well as I have demonstrated in the C++ code of issue #116583.
There I overrun the inalloca stack object within the called function itself.
This is actually something I find more often in my test code base then "alloca" calls.

Lifetime annoations and stacksave/restore intrinsics will get properly inlined into a caller if they exist.
My initial thought was to then maybe allow WinEH to process the lifetimes of these dynamically allocated stack objects.
The problem that I encountered with this approach is the following:

```llvm
; Function Attrs: mustprogress sspstrong
define internal fastcc noundef i32 @"??$invoke@AAV<lambda_0>@?0??..."(ptr noundef nonnull align 4 dereferenceable(48) %_Obj, ptr noundef nonnull align 4 dereferenceable(8) %_Arg1, ptr noundef nonnull align 4 dereferenceable(4) %_Args21, ptr noundef nonnull align 4 dereferenceable(28) %_Args2) unnamed_addr #3 !dbg !7302 {
entry:
  tail call void @llvm.dbg.value(metadata ptr %_Args2, metadata !7306, metadata !DIExpression()), !dbg !7315
  tail call void @llvm.dbg.value(metadata ptr %_Args21, metadata !7307, metadata !DIExpression()), !dbg !7315
  tail call void @llvm.dbg.value(metadata ptr %_Arg1, metadata !7308, metadata !DIExpression()), !dbg !7315
  tail call void @llvm.dbg.value(metadata ptr %_Obj, metadata !7309, metadata !DIExpression()), !dbg !7315
  %argmem = alloca inalloca <{ %"class.std::shared_ptr.16", ptr, ptr }>, align 4
  %call = call x86_thiscallcc noundef ptr @"??0?$shared_ptr@..."(ptr noundef nonnull align 4 dereferenceable(8) %argmem, ptr noundef nonnull align 4 dereferenceable(8) %_Arg1) #18, !dbg !7316
  %0 = getelementptr inbounds <{ %"class.std::shared_ptr.16", ptr, ptr }>, ptr %argmem, i32 0, i32 1, !dbg !7316
  store ptr %_Args21, ptr %0, align 4, !dbg !7316
  %1 = getelementptr inbounds <{ %"class.std::shared_ptr.16", ptr, ptr }>, ptr %argmem, i32 0, i32 2, !dbg !7316
  store ptr %_Args2, ptr %1, align 4, !dbg !7316
  %call3 = call fastcc noundef i32 @"??R<lambda_0>@?0??.."(ptr noundef nonnull %_Obj, ptr nonnull %argmem), !dbg !7316, !range !7232
  ret i32 %call3, !dbg !7316
}
```

Here we have a std::invoke instantiation (truncated for readability) which does execute the call operator of the lambda.
We do not have lifetimes for the argmem and also the call operator has its inalloca attribute stripped.

I see the case for not emitting lifetime annotations here in the frontend due to this being a single block func.
But what is more problematic in my opinion is that GlobalOpt does strip the inalloca attribute of argmem inside the CallBase but doesnt strip the attribute for the alloca.
If we now inline the call operator func and want to generate lifetimes for the inalloca alloca we would have to walk all instructions to check if the caller has an inalloca alloca.
We can't determine it from the CallBase since the Inliner runs after GlobalOpt.


So TL;DR is: 
I guess the lifetime management of inalloca allocas is also suboptimal and could be improved.

Comment (MuellerMP) [reply llvm/lib/Target/X86/X86WinEHState.cpp]: I wonder if there is any security aspect to this since the original bug was causing a stack object overrun, which could theoretically be used for exploitation in some weird scenario.
But since this is only caused by a miscompile of clang-cl I'm uncertain wheter this is actually relevant.

Comment (rnk) [reply llvm/lib/Target/X86/X86WinEHState.cpp]: > But what is more problematic in my opinion is that GlobalOpt does strip the inalloca attribute of argmem inside the CallBase but doesnt strip the attribute for the alloca.

That is a good point!

I would lean away from messing with lifetime attribute placement. There are many aspects of exception handling and optimizations that make it hard to use these reliably.

A low-cost idea that comes to mind is to have a local pass that looks for inalloca allocas that are not used in conjunction with an inalloca call and promote them to static allocas. There are many passes beyond inlining that might delete the consuming inalloca call, so there are good reasons to do this separately from GlobalOpt. However, inalloca allocas are often captured by constructors and destructors, which is why we're using inalloca in the first place, and that makes it hard to prove the alloca is not used later by an inalloca call.

I previously attempted to design a more complete solution to this problem with the `preallocated` attribute which uses tokens to make it possible to statically identify which calls use the argument memory, but the reality is that it's not a priority for my employer, so I wasn't able to get this work prioritized.

Comment (rnk) [reply llvm/lib/Target/X86/X86WinEHState.cpp]: Regarding the security implications, I would just fix the miscompile. The on-stack exception handling records make this platform extremely easy to attack, and I can't say with confidence that LLVM has implemented all of the hardening techniques to protect the exception registration nodes in the FS:00 linked list.

Comment (MuellerMP) [reply llvm/lib/Target/X86/X86WinEHState.cpp]: Thanks for your insight @rnk - greatly appreciated!
Fixing up callless inalloca allocas will probably reduce the impact this fix has on O2.
I'm going to look into it the following couple of weekends.
I guess my current fix is still valid for (atleast) those edge cases of using dynamic allocas or statement expressions.
Maybe I should separate the continued work on hoisting callless inalloca allocas if you have concerns regarding the reliability of proving them (new PR).

PS: Not to be obnoxious but if you find any more spare time to spend on WinEH reviews - a colleague of mine has prepped some state handling improvements for /EHa over at #116546.

**Review by rnk — APPROVED** [review]
Thanks for the fix! I'll merge it
